Fix attempt to free static OOM error message

mhp opened the pull request:
I noticed a mechanism that can lead to attempts to free the statically allocated OOM error message (whilst trying to reproduce #3313).  The function `giterr_detach` assumes that any error message it detaches is dynamically allocated.  This PR fixes that, so that it becomes safe to detach an OOM error.

I have also added a test that performs the full `giterr_capture`/`giterr_restore` cycle, checking to ensure that the static pointer is not saved in the `git_error_state` structure.

Comments welcome!

ethomson commented:
Yes, this is definitely problematic.  I noticed similar problems in #3257, though it's not appropriate for inclusion here.  I think there may be some possibility for leaking here also since we don't provide a way to free the detached buffer.

In any case, if you're hitting an OOM then I don't like to assume that a `strdup` will succeed.  Can we return the buffer and test to make sure that we don't try to free the static OOM string?  Or if the `strdup` fails, are we left in an okay place?

mhp commented:
I don't think I've increased the leak risk - as far as I can see, the detached buffer is expected to be dynamically allocated.  I didn't see a disposal function anywhere (`giterr_free()` perhaps?), so I assume users of this function know to dispose of the buffer themselves.

You are right that `strdup` seems likely to fail under oom conditions.  I believe that this would end up storing a null pointer in the structure.  This isn't ideal, but the `klass` is still valid for diagnostics to some degree.

Returning the static buffer pointer, and checking when freeing would probably work very well for the `giterr_capture`/`giterr_restore` cycle, but I don't think it would work if just the `giterr_detach` function is used, simply because we don't have a free-like function to handle the disposal side.

Have I misinterpreted the semantics of the `giterr_detach` function?  If there is an approved disposal route, we can add the pointer matching there to make it all work.

ethomson commented:
No, I don't think you've increased leakiness either, I was just pointing out that we're leaky now.

My opinion is that [`giterr_detach`](`https://github.com/libgit2/libgit2/blob/master/include/git2/errors.h#L125) shouldn't be public at all.  We offer callers the error information, I don't see the value in a single call that gets and clears it, with no way for the caller to free that information safely.

To clarify:  you're fixing this because the library is buggy and ended up thrashing your app, correct?  (That is to say, you don't actually want to call `giterr_detach` from your code, and you wouldn't mind seeing it disappear?)

mhp commented:
That's right - I'm not interested in `giterr_detach`.  I'd like to be able to `giterr_capture`/`giterr_restore` an OOM condition without freeing static data though!

Would you like me to rework this so that `giterr_detach` is no longer public and the static message is handled correctly without attempting to `strdup` it?

mhp commented:
I've removed `giterr_detach` from the public API, and fixed `giterr_restore` to notice if an OOM condition is being restored.  I think this fixes the problems I saw initially (attempt to free static memory) without resorting to `strdup`!

What do you think, @ethomson?  I could also move the body of `giterr_detach` into `giterr_capture` if you think this is the right direction to move in.
